### PR TITLE
Collapse help into expandable command topics

### LINKED ISSUE #1688: dokku --help is unwieldy, collapse help into commands (topics) similar to heroku and docker
Presently dokku outputs some 60+ lines of help with no paging option built in, before you add plugins. When hunting for a command, a user may not be familiar with the naming conventions and be left to hunt for the feature.

Example of output today:
```
$ dokku --help
Usage: dokku [--quiet|--trace|--rm-container|--rm|--force] COMMAND <app> [command-specific-options]

Options:
    apps:create <app>                                                  Create a new app
    apps:destroy <app>                                                 Permanently destroy an app
    apps                                                               List your apps
    certs:add <app> CRT KEY                                            Add an ssl endpoint to an app. Can also import from a tarball on stdin
    certs:chain CRT [CRT ...]                                          [NOT IMPLEMENTED] Print the ordered and complete chain for the given certificate.
    certs:generate <app> DOMAIN                                        Generate a key and certificate signing request (and self-signed certificate)
    certs:info <app>                                                   Show certificate information for an ssl endpoint.
    certs:key <app> CRT KEY [KEY ...]                                  [NOT IMPLEMENTED] Print the correct key for the given certificate.
    certs:remove <app>                                                 Remove an SSL Endpoint from an app.
    certs:rollback <app>                                               [NOT IMPLEMENTED] Rollback an SSL Endpoint for an app.
    certs:update <app> CRT KEY                                         Update an SSL Endpoint on an app. Can also import from a tarball on stdin
    config (<app>|--global)                                            Display all global or app-specific config vars
    config:get (<app>|--global) KEY                                    Display a global or app-specific config value
    config:set (<app>|--global) KEY1=VALUE1 [KEY2=VALUE2 ...]          Set one or more config vars
    config:unset (<app>|--global) KEY1 [KEY2 ...]                      Unset one or more config vars
    docker-options:add <app> <phase(s)> OPTION                         Add docker option to app for phase (comma separated phase list)
    docker-options <app>                                               Display apps docker options for all phases
    docker-options <app> <phase(s)>                                    Display apps docker options for phase (comma separated phase list)
    docker-options:remove <app> <phase(s)> OPTION                      Remove docker option from app for phase (comma separated phase list)
    domains:add <app> DOMAIN                                           Add a custom domain to app
    domains <app>                                                      List custom domains for app
    domains:clear <app>                                                Clear all custom domains for app
    domains:remove <app> DOMAIN                                        Remove a custom domain from app
    enter <app>  [<container-type> || --container-id <container-id>]   Connect to a specific app container
    events:list                                                        List logged events
    events:off                                                         Disable events logger
    events:on                                                          Enable events logger
    events [-t]                                                        Show the last events (-t follows)
    help                                                               Print the list of commands
    logs <app> [-h] [-t] [-n num] [-q] [-p process]                    Show the last logs for an application
    ls                                                                 Pretty listing of deployed applications and containers
    nginx:access-logs <app> [-t]                                       Show the nginx access logs for an application (-t follows)
    nginx:build-config <app>                                           (Re)builds nginx config for given app
    nginx:disable <app>                                                disable nginx for an application (forces container binding to external interface)
    nginx:enable <app>                                                 enable nginx for an application
    nginx:error-logs <app> [-t]                                        Show the nginx error logs for an application (-t follows)
    plugin:disable <name>                                              Disable an installed plugin (third-party only)
    plugin:enable <name>                                               Enable a previously disabled plugin
    plugin:install [--core|git-url]                                    Optionally download git-url & run install trigger for active plugins (or only core ones)
    plugin:install-dependencies [--core]                               Run install-dependencies trigger for active plugins (or only core ones)
    plugin                                                             Print active plugins
    plugin:uninstall <name>                                            Uninstall a plugin (third-party only)
    plugin:update [git-url]                                            Optionally download git-url & run update trigger for active plugins
    ps <app>                                                           List processes running in app container(s)
    ps:rebuildall                                                      Rebuild all apps
    ps:rebuild <app>                                                   Rebuild an app
    ps:restartall                                                      Restart all deployed app containers
    ps:restart <app>                                                   Restart app container(s)
    ps:restore                                                         Start previously running apps e.g. after reboot
    ps:scale [<app> <proc>=<count> [<proc>=<count>]]                   Get/Set how many instances of a given process to run
    ps:start <app>                                                     Start app container(s)
    ps:stop <app>                                                      Stop app container(s)
    run <app> <cmd>                                                    Run a command in the environment of an application
    shell                                                              Spawn dokku shell
    tags <app>                                                         List all app image tags
    tags:create <app> <tag>                                            Add tag to latest running app image
    tags:deploy <app> <tag>                                            Deploy tagged app image
    tags:destroy <app> <tag>                                           Remove app image tag
    tar:from <app> <url>                                               Loads an app tarball from url.
    tar:in <app>                                                       Reads an tarball containing the app from stdin
    trace [on/off]                                                     Enable dokku tracing
    url <app>                                                          Show the first URL for an application (compatibility)
    urls <app>                                                         Show all URLs for an application
    version                                                            Print dokku's version
```

***Proposed change:***
```
$ dokku --help
Usage: dokku [--quiet|--trace|--rm-container|--rm|--force] COMMAND <app> [command-specific-options]

Help commands, use 'dokku [COMMAND] --help' for more details:

    apps                                                               List your apps
    certs                                                              Update an SSL Endpoint on an app. Can also import from a tarball on stdin
    config                                                             Display all global or app-specific config vars
    docker-options                                                     Display apps docker options for all phases
    domains                                                            List custom domains for app 
    enter                                                              Connect to a specific app container
    events                                                             Show the last events (-t follows)
    logs                                                               Show the last logs for an application
    ls                                                                 Pretty listing of deployed applications and containers
    plugin                                                             Print active plugins
    ps                                                                  List processes running in app container(s)
    run                                                                Run a command in the environment of an application
    shell                                                              Spawn dokku shell
    tags                                                               List all app image tags
    tar                                                                Reads an tarball containing the app from stdin

Additional commands:
    help                                                               Print the list of commands
    trace [on/off]                                                     Enable dokku tracing
    url <app>                                                          Show the first URL for an application (compatibility)
    urls <app>                                                         Show all URLs for an application
    version                                                            Print dokku's version

$ dokku ps --help

Usage: dokku ps[:command-specific-options] [<app>] 

    ps <app>                                                           List processes running in app container(s)
    ps:rebuildall                                                      Rebuild all apps
    ps:rebuild <app>                                                   Rebuild an app
    ps:restartall                                                      Restart all deployed app containers
    ps:restart <app>                                                   Restart app container(s)
    ps:restore                                                         Start previously running apps e.g. after reboot
```

Formating above is just an example and would be corrected for description and posix utility argument syntax conventions. This is obviously a `plugn` issue, but since dokku is the main consumer of that project, I figured I would start the conversation here before I submitted a PR there.

### PR DESCRIPTION
**Pull request by u2mejc**
This resolves #1688 :+1: 

```
$ dokku
Usage: dokku [--quiet|--trace|--rm-container|--rm|--force] COMMAND <app> [command-specific-options]

Primary help options, type "dokku COMMAND:help" for more details, or dokku help --all to see all commands.
Commands:
    apps             List your apps
    checks           Show zero-downtime status
    config           Display all global or app-specific config vars
    docker-options   Display app's Docker options for all phases (or comma separated phase list)
    domains          List domains
    enter            Connect to a specific app container
    events           Show the last events (-t follows)
    logs             Show the last logs for an application
    ls               Pretty listing of deployed applications and containers
    plugin           Print active plugins
    proxy            Show proxy for app
    run              Run a command in the environment of an application
    tags             List all app image tags
    trace            Enable dokku tracing
    url              Show the first URL for an application (compatibility)
    urls             Show all URLs for an application
    version          Print dokku's version

$ dokku domains:help
    domains:add <app> DOMAIN      Add a domain to app
    domains [<app>]               List domains
    domains:clear <app>           Clear all domains for app
    domains:disable <app>         Disable VHOST support
    domains:enable <app>          Enable VHOST support
    domains:remove <app> DOMAIN   Remove a domain from app
    domains:set-global <domain>   Set global domain name
```

### CODE REVIEW

**Comment by michaelshobbs** [comment]
Can you show what the outputs look like in the description of the PR?

File: dokku [hunk 1]
```diff
@@ -159,11 +159,16 @@ case "$1" in
   help|'')
     echo "Usage: dokku [--quiet|--trace|--rm-container|--rm|--force] COMMAND <app> [command-specific-options]"
     echo ""
-    echo "Options:"
+    echo "Primary help options, type \"dokku COMMAND:help\" for more details, or dokku help --all to see all commands."
+    echo "Commands:"
```
Comment (michaelshobbs) [inline dokku]: I think consolidating lines 162 and 163 to the following (which is an almost direct copy of heroku's msg) makes things clearer.
```
    echo "Primary help topics, type 'dokku TOPIC:help' for more details:"
```

Comment (u2mejc) [inline dokku]: You've obscured the --all, I added that feature for you. Please advise. :wink:

**Comment by michaelshobbs** [comment]
@u2mejc Also, I have thoughts on how to do this if you'd like to discuss offline.

File: plugins/docker-options/commands [hunk 2]
```diff
@@ -4,12 +4,15 @@ set -eo pipefail; [[ $DOKKU_TRACE ]] && set -x
 
 case "$1" in
   help | docker-options:help)
-    cat<<EOF
-    docker-options <app>, Display app's Docker options for all phases
-    docker-options <app> <phase(s)>, Display app's Docker options for phase (comma separated phase list)
+    help_content='    docker-options <app>, Display app'"'"'s Docker options for all phases
+    docker-options <app> <phase(s)>, Display app'"'"'s Docker options for phase (comma separated phase list)
     docker-options:add <app> <phase(s)> OPTION, Add Docker option to app for phase (comma separated phase list)
```
Comment (michaelshobbs) [inline plugins/docker-options/commands]: That's not a helpful response. Can you clarify please?

**Comment by u2mejc** [comment]
> Any way to get the bare method (domains in this case) above other subcommands? Just looks stupid atm.

Was discussed here: https://github.com/dokku/dokku/pull/1830 
> michaelshobbs commented 8 days ago
> It's not that simple since some plugin commands do something functional when no args are given.

**Comment by michaelshobbs** [comment]
I don't think that's the same context. I believe @josegonzalez is asking about sort order of the sub topic output. i.e.

note that the bare `dokku domains` command is first.
```
$ dokku domains:help
    domains [<app>]               List domains
    domains:add <app> DOMAIN      Add a domain to app
    domains:clear <app>           Clear all domains for app
    domains:disable <app>         Disable VHOST support
    domains:enable <app>          Enable VHOST support
    domains:remove <app> DOMAIN   Remove a domain from app
    domains:set-global <domain>   Set global domain name
```

**Comment by josegonzalez** [comment]
@michaelshobbs that "additional commands" section on the bare commands could break parsing output.

**Comment by michaelshobbs** [comment]
Parsing how? Does help output need to be machine readable?

**Comment by josegonzalez** [comment]
No but we can't expect to parse:

```
dokku apps
 === My Apps
 example
 example2

 === Collaborated Apps
 theirapp   other@owner.name

Additional commands, type "heroku help COMMAND" for more details:

  apps:create [NAME]             #  create a new app
  apps:destroy --app APP         #  permanently destroy an app
  apps:info                      #  show detailed app information
  apps:join --app APP            #  add yourself to an organization app
  apps:leave --app APP           #  remove yourself from an organization app
  apps:lock --app APP            #  lock an organization app to restrict access
  apps:open --app APP            #  open the app in a web browser
  apps:rename NEWNAME --app APP  #  rename the app
  apps:transfer RECIPIENT        #  transfer an app to another user or organization
  apps:unlock --app APP          #  unlock an organization app so that any org member can join it
```

**Comment by josegonzalez** [comment]
I just don't want the extra help output on things like `dokku apps`. I'm agreeing with everything else.

**Comment by michaelshobbs** [comment]
Oh I'm not saying the `dokku apps` command should output help. Instead `dokku apps:help` would output this screen.

**Comment by josegonzalez** [comment]
Ah okay, thats fine by me :)

File: plugins/docker-options/commands (hunk 2 above)
Comment (u2mejc) [inline plugins/docker-options/commands]: ah sorry about that, it's because that's a separate issue about POSIX argument syntax standards compliance. We have something of a history of taking things into the weeds on minor points like that and I was hoping to not touch the topic (since that was preexisting before this PR).

Comment (michaelshobbs) [inline plugins/docker-options/commands]: I think it's fine to refactor this bit in the context of cleaner help output.

**Comment by michaelshobbs** [comment]
**Top level help output**
```
Usage: dokku [--quiet|--trace|--rm-container|--rm|--force|--help] COMMAND <app> [command-specific-options]

Primary help topics, type "dokku TOPIC:help" for more details:
    apps                # manage apps (create, destroy)
    checks              # manage zero downtime features
    config              # manage global or app-specific configuration
    docker-options      # manage docker options for all phases
    domains             # manage domains
    enter               # connect to a specific app container
    logs                # display logs for an app
    proxy               # manage proxy configuration
    run                 # run a command in the environment of an app
    tags                # manage app image tags

Additional topics:
    events              # display events
    ls                  # display deployed apps and containers
    plugin              # manage plugins
    trace               # enable debug tracing
    version             # display version
    urls                # display all URLs for an app

3rd party plugin topics:
  deploymentkeys:create <app>   Create a pair of app-specific deployment keys
  deploymentkeys:delete <app>   Delete the current pair of deployment keys
  deploymentkeys:shared         Shows the current shared public key
  deploymentkeys:show   <app>   Shows the current public key to add to your VCS
```

**Core topic detail**
```
Usage: dokku apps

 list your apps

Example:

 $ dokku apps
=====> My Apps
 example
 example2

Additional commands:

  apps                              # list your apps
  apps:create <app>                 # create a new app
  apps:destroy <app>                # permanently destroy an app
  apps:rename <old-app> <new-app>   # rename an app
```

Notes
- some of the top level copy has changed
- there is a new 3rd party section for non-core help output
- we'll want to come up with a clean API to give to 3rd party devs for help output (something like define a usage() function with a given text format and we'll handle the collation)
 
I'm sure I'm missing something here so feedback is welcome

**Comment by josegonzalez** [comment]
I don't want to do this:

```
 $ dokku apps
=====> My Apps
 example
 example2

Additional commands:

  apps                              # list your apps
  apps:create <app>                 # create a new app
  apps:destroy <app>                # permanently destroy an app
  apps:rename <old-app> <new-app>   # rename an app
```

For bare commands that already have output, adding more things to the end seems... wrong?

**Comment by michaelshobbs** [comment]
Ok were confused again. I'm not saying to change the functionality of `dokku apps`. This is only the output of `dokku apps:help`

**Comment by josegonzalez** [comment]
Yeah but the command you have up there is `dokku apps` and you're appending `dokku apps:help` to it. I don't want to append the help output to commands that already do something.

**Comment by michaelshobbs** [comment]
I'm not following. 

`dokku apps` does what it always did. Help will not be shown in any other case than `dokku TOPIC:help`

**Comment by josegonzalez** [comment]
Ah shit, I didn't realize that whole block was the output of `dokku app:help`. Fuck, my bad.

**Comment by u2mejc** [comment]
How do you guys feel about me drawing a hardline between "core" and other plugins in the file structure? In the Makefile (haven't looked at the deb) we symlink core plugins to both directories which is breaking my devision here. Cool if we respect the divide as: if it's in this repo, it's core, if not, then it's not core?

**Comment by michaelshobbs** [comment]
My thinking was that core plugins are those in `/var/lib/dokku/core-plugins`. I don't think that includes official plugins like MySQL, PostgreSQL, etc. Is that true? 😁

**Comment by josegonzalez** [comment]
Fine by me. @michaelshobbs did similar with the `--app` flag in the commands refactor.

**Comment by josegonzalez** [comment]
Correct, those are official, but not core.

**Comment by u2mejc** [comment]
cool, we're on the same page.

**Comment by u2mejc** [comment]
I split these into two commits, and I think it shouldn't be squashed, but we can move to another PR if needed.

This follows up the work of dividing the "community" (don't say third party) plugin paths. Specifically, when the Makefile was updated to no long insert symlinks in /var/lib/dokku/plugin, it removed the assumption that all plugins could be in either dir. After following some other avenues, I decided to simply wrap plugn in a function (since it relies on a global var).

File: plugins/ps/commands [hunk 3]
```diff
@@ -4,8 +4,7 @@ set -eo pipefail; [[ $DOKKU_TRACE ]] && set -x
 
 case "$1" in
   help | ps:help)
-    cat<<EOF
-    ps <app>, List processes running in app container(s)
+    help_contet='    ps <app>, List processes running in app container(s)
```
Comment (michaelshobbs) [inline plugins/ps/commands]: typo?

File: plugins/shell/commands [hunk 4]
```diff
@@ -4,9 +4,13 @@ set -eo pipefail; [[ $DOKKU_TRACE ]] && set -x
 
 case "$1" in
   help | shell:help)
-    cat<<EOF
-    shell, Spawn dokku shell
-EOF
+    help_contnet='    shell, Spawn dokku shell'
```
Comment (michaelshobbs) [inline plugins/shell/commands]: typo?

File: Makefile [hunk 5]
```diff
@@ -57,10 +57,8 @@ copyfiles:
 		rm -rf ${CORE_PLUGINS_PATH}/$$plugin && \
 		rm -rf ${PLUGINS_PATH}/$$plugin && \
 		cp -R plugins/$$plugin ${CORE_PLUGINS_PATH}/available && \
-		ln -s ${CORE_PLUGINS_PATH}/available/$$plugin ${PLUGINS_PATH}/available; \
 		find /var/lib/dokku/ -xtype l -delete;\
 		PLUGIN_PATH=${CORE_PLUGINS_PATH} plugn enable $$plugin ;\
-		PLUGIN_PATH=${PLUGINS_PATH} plugn enable $$plugin ;\
```
Comment (michaelshobbs) [inline Makefile]: I seem to be missing where we enable plugins on install. Can you point me there? I **think** this is also executed in the the deb package.

**Comment by michaelshobbs** [comment]
Can we make the split between community and core output without wrapping the plugn execution, and changing the plugin path variables?

File: plugins/git/commands [hunk 6]
```diff
@@ -54,11 +54,11 @@ git_hook_cmd() {
     if [[ $refname = "refs/heads/master" ]]; then
       # broken out into plugin so we might support other methods to receive an app
       # shellcheck disable=SC2086
-      plugn trigger receive-app $APP $newrev
+      trigger-plugn trigger receive-app $APP $newrev
     else
-      if [[ $(find "$PLUGIN_PATH"/enabled/*/receive-branch 2>/dev/null | wc -l) != 0 ]]; then
+      if [[ $(find "$PLUGIN_CORE_PATH"/enabled/*/receive-branch 2>/dev/null | wc -l) != 0 || $(find "$PLUGIN_COMMUNITY_PATH"/enabled/*/receive-branch 2>/dev/null | wc -l) != 0 ]]; then
```
Comment (michaelshobbs) [inline plugins/git/commands]: Probably could do this with one conditional like this:
```
if [[ $(find "$DOKKU_LIB_ROOT"/enabled/*/receive-branch 2>/dev/null | wc -l) != 0 ]]; then
```

**Comment by michaelshobbs** [comment]
Example of how to only get core plugin help
```
    CORE_PLUGIN_COMMANDS=$(find -L "$PLUGIN_CORE_ENABLED_PATH" -name commands 2> /dev/null || true)
    for plugin_command in $CORE_PLUGIN_COMMANDS; do
      $plugin_command help
    done | sort | column -c2 -t -s,
```

**Comment by u2mejc** [comment]
I totally understand and I even created code that was similar originally. The problem is that the rest of dokku expects to find core plugins in /var/lib/dokku/plugins/.

If this wasn't related to the help PR, how would you feel about segregating the core / community plugins and remove coding calling plugn directly?

**Comment by michaelshobbs** [comment]
The PR pedantics aside, I don't see how the added complexity is beneficial.

**Comment by michaelshobbs** [comment]
I'm not following what's wrong with your original approach and similarly my code example. Can you elaborate?

**Comment by u2mejc** [comment]
Totally, and I agree, complexity for the sake of complexity doesn't make sense. 

Today, the core plugins (plugins that are part of this repo) are located in `/var/lib/dokku/core-plugins/available/`. Both the deb and the Makefile then symlink the core plugins to `/var/lib/dokku/core-plugins/enabled/` using `plugn enabled`, but not before the script symlinks the contents of `/var/lib/dokku/core-plugins/available` to `/var/lib/dokku/plugins/available` and then enabling them with `plugn enable` in both places. Take a look at this code block from debin/postinst for context:
```
    echo "Enabling all core plugins"
    find ${DOKKU_LIB_ROOT}/core-plugins/available -mindepth 1 -maxdepth 1 -type d -printf '%f\n' | while read -r plugin; do
      if [ ! -d ${DOKKU_LIB_ROOT}/plugins/available/$plugin ]; then
        ln -s ${DOKKU_LIB_ROOT}/core-plugins/available/$plugin ${DOKKU_LIB_ROOT}/plugins/available/$plugin;
        PLUGIN_PATH=${DOKKU_LIB_ROOT}/core-plugins plugn enable $plugin
        PLUGIN_PATH=${DOKKU_LIB_ROOT}/plugins plugn enable $plugin
      fi
    done
```

Related, in the Makefile, anytime `make copyfiles` runs, including install, all of the user's "community" plugins are spontaneously enabled. I've seen this has the potential to leave dokku in a broken state if a community plugin is broken.

The reason for this current state of file hierarchy may become apparent when looking at the code, parts of Dokku still expect all plugins (core or otherwise) to be found in /var/lib/dokku/plugins/ today. Both new world and old world are intertangled commonly like this:
```
source "$PLUGIN_CORE_AVAILABLE_PATH/common/functions"
source "$PLUGIN_AVAILABLE_PATH/checks/functions"
```
 Because of this assumption, it is difficult to make a clean division from what is core and what is community. Reading the symlink path like in the code example, would be one way to skin this cat. But it increases the glue code throughout the code base. We could just wrap up that glue in a common function to return if a passed plugin is core or not for each call, but we don't resolve for the human condition. Discerning sys admins may look at the directory system themselves, keeping that clean divide by not symlinking core plugins into the same directory as community plugins will help in avoiding the confusion of human error.

**Comment by michaelshobbs** [comment]
Understood. I don't want to get too much in to the history of this decision but see the following for context
#1205 #1425 #1228 
To be clear, I'm also not debating that this PR is functional. 

`plugn` requires `PLUGIN_PATH` to be set in order to execute properly; as you know. In this PR we are now forced to run `plugn` twice; once for each set of plugins. How is this additional complexity justified? More to your point about increased "glue code throughout the code base", the symlink check only needs to be used when calling help. Admittedly we also do something similar in `execute_dokku_cmd()`. To that point, we could make this a common function in the future. (i.e. get_core_plugins or something like that.)  

I strongly think the symlink check instead of the reorganization (yet again) of the plugin directory structure is the most efficient and effective path to making this particular change.

**Comment by josegonzalez** [comment]
I defer to @michaelshobbs on this pr.

File: plugins/00_dokku-standard/subcommands/deploy [hunk 7]
```diff
@@ -1,8 +1,8 @@
 #!/usr/bin/env bash
 set -eo pipefail; [[ $DOKKU_TRACE ]] && set -x
 source "$PLUGIN_CORE_AVAILABLE_PATH/common/functions"
-source "$PLUGIN_AVAILABLE_PATH/checks/functions"
-source "$PLUGIN_AVAILABLE_PATH/config/functions"
+source "$PLUGIN_CORE_AVAILABLE_PATH/checks/functions"
+source "$PLUGIN_CORE_AVAILABLE_PATH/config/functions"
```
Comment (josegonzalez) [inline plugins/00_dokku-standard/subcommands/deploy]: Why these changes? We could previously override the implementations of core plugins, and now we can't.

File: docs/docker-options.md [hunk 8]
```diff
@@ -5,8 +5,7 @@
 Pass [options](https://docs.docker.com/engine/reference/run/) to Docker during Dokku's `build`, `deploy` and `run` phases
 
 ```
-docker-options <app>                             Display app's Docker options for all phases
-docker-options <app> <phase(s)>                  Display app's Docker options for phase (comma-separated phase list)
+docker-options <app> [phase(s)], Display app'"'"'s Docker options for all phases (or comma separated phase list)
```
Comment (josegonzalez) [inline docs/docker-options.md]: Mind adding the proper spacing instead?

**Comment by u2mejc** [comment]
Refactored to drop the plugn wrapper and utilize most of @michaelshobbs example.

> Mind adding the proper spacing instead?

Done!

**Comment by josegonzalez** [comment]
What about the change in sourcing for core plugins?

**Comment by u2mejc** [comment]
@josegonzalez asked:
> What about the change in sourcing for core plugins?

After talking to Michael, we agreed that commit exceeded the scope of this "help" PR. I dropped that commit and just updated the help command to check it's symlink path.

If we're interested, we should probably open a separate RFC for plugin paths, managing plugins, and overriding / disabling core plugins.

File: dokku [hunk 9]
```diff
@@ -157,17 +157,49 @@ execute_dokku_cmd() {
...
+          $community_plugin_command help
+        done | sort | column -c2 -t -s,
+      fi
+    fi
+  ;;
 
   *)
     execute_dokku_cmd "$@"
-    ;;
+  ;;
```
Comment (josegonzalez) [inline dokku]: :(

File: dokku [hunk 10]
```diff
@@ -157,17 +157,49 @@ execute_dokku_cmd() {
...
+        $core_plugin_command help
+      done | sort | sed -e '/^.*:/d' -e 's/\s[\[\<\-\(].*,/,/' | column -c2 -t -s,
+      if [[ -n "$COMMUNITY_PLUGIN_COMMANDS" ]]; then
+        echo -e "\nCommunity plugin commands:\n"
+        for community_plugin_command in $COMMUNITY_PLUGIN_COMMANDS; do
+          $community_plugin_command help
+        done | sort | column -c2 -t -s,
+      fi
+    fi
+  ;;
```
Comment (josegonzalez) [inline dokku]: :( :(

**Comment by josegonzalez** [comment]
Last two comments, and then I'm :+1: on this PR.

**Comment by michaelshobbs** [comment]
Probably should update the plugin development docs to show how one might use the new help format

**Comment by u2mejc** [comment]
will do, one sec

**Comment by u2mejc** [comment]
Any other requests before we :shipit: ?

File: dokku [hunk 11]
```diff
@@ -157,13 +157,45 @@ execute_dokku_cmd() {
...
+        $core_plugin_command help
+      done | sort | column -c2 -t -s,
+      echo -e "\nCommunity plugin commands:\n"
+      for community_plugin_command in $COMMUNITY_PLUGIN_COMMANDS; do
+        $community_plugin_command help
+      done | sort | column -c2 -t -s,
+    else
+      for core_plugin_command in $CORE_PLUGIN_COMMANDS; do
+        $core_plugin_command help
+      done | sort | sed -e '/^.*:/d' -e 's/\s[\[\<\-\(].*,/,/' | column -c2 -t -s,
```
Comment (michaelshobbs) [inline dokku]: Can you add a comment on what this `sed` regex is doing?

**Comment by michaelshobbs** [comment]
Couple thoughts:
- I'd like to see some more descriptive documentation around the `help_content` variable and how it's used. Maybe mention in 0.5.x we expanded the usability of help and here's why you want to use it. Include a code block that shows how a user interfaces with `help` and what output looks like when you do and don't use `help_content`. 
- I'd prefer to avoid `\n` and instead just use multiple `echo` statements. I think this makes our core usage more "accessible" to community plugin devs as examples for implementation.

LMK if this doesn't make any sense

File: dokku (hunk 11 above)
Comment (u2mejc) [inline dokku]: Can do. 

Of note, I've intentionally been using less comments because the rest of the project has been using less comments. We use some declares to bubble up functions but very little comments anywhere else to explain code (especially would be nice for weird things like bash substitutions). If we're worried about bloat since the project is script based, we could have the comments stripped in the releases and only visible when checked out.

Comment (michaelshobbs) [inline dokku]: Fair point around comments. My personal opinion on them is to only use them to describe functionality that isn't blatantly apparent. Is some cases we've wrapped the functionality in a "well-named" function. Dokku's consistency, aside I think it makes sense to use a comment here for now.

**Comment by u2mejc** [comment]
_michaelshobbs commented 7 minutes ago_
>I'd like to see some more descriptive documentation around the help_content variable and how it's used. Maybe mention in 0.5.x we expanded the usability of help and here's why you want to use it. Include a code block that shows how a user interfaces with help and what output looks like when you do and don't use help_content.

Understood.

>I'd prefer to avoid \n and instead just use multiple echo statements. I think this makes our core usage more "accessible" to community plugin devs as examples for implementation.
LMK if this doesn't make any sense

Your reason is sound and I'll conform, but it doesn't mean I have to like it. :kissing_heart:

**Comment by u2mejc** [comment]
Anything else after those? :smiley:

**Comment by josegonzalez** [comment]
Nothing from me.

File: docs/docker-options.md [hunk 12]
```diff
@@ -5,8 +5,7 @@
 Pass [options](https://docs.docker.com/engine/reference/run/) to Docker during Dokku's `build`, `deploy` and `run` phases
 
 ```
-docker-options <app>                             Display app's Docker options for all phases
-docker-options <app> <phase(s)>                  Display app's Docker options for phase (comma-separated phase list)
+docker-options <app> [phase(s)]                  Display app's Docker options for all phases (or comma separated phase list)
```
Comment (michaelshobbs) [inline docs/docker-options.md]: +1 for POSIX standards :smile:

**Comment by michaelshobbs** [comment]
Oh! My last bit would be to move the help commands to functions that include all local vars (or at least as much as possible); including `desc` and `cmd` vars like we do in other areas of the core plugin suite.

**Comment by u2mejc** [comment]
@michaelshobbs When I need to edit multiple files, I use `vim $(grep blah blah)` or even `sed -i`. Your symlinks to files I've already edited in the repo are screwing with my flow. :cry:

**Comment by michaelshobbs** [comment]
Feel free to use a "real" editor that understand symlinks or open an RFC issue to discuss. :smile:

**Comment by josegonzalez** [comment]
Nano is pretty good.

**Comment by michaelshobbs** [comment]
:trollface:

File: docs/development/plugin-triggers.md [hunk 13]
```diff
@@ -92,10 +92,10 @@ cd /root/dokku
 sudo BUILD_STACK=true make install
 ```
 
-### `commands help`
+### `comands help` and `commands <PLUGIN_NAME>:help`
 
-- Description: Used to aggregate all plugin `help` output. Your plugin should implement a `help` command in your `commands` file to take advantage of this plugin trigger. This must be implemented inside the `commands` plugin file.
-- Invoked by: `dokku help`
+- Description: Your plugin should implement a `help` command in your `commands` file to take advantage of this plugin trigger. `commands help` is used by `dokku help` to aggregate all plugins abbreviated `help` output. Implementing  `<PLUGIN_NAME>:help` in your `commands` file gives users looking for help, a more detail output. 'commands help' must be implemented inside the `commands` plugin file. It's recommended that PLUGIN_NAME:help be added to the commands file to ensure consitancty among community plugins and give you a new avenue to share rich help content with your user.
```
Comment (michaelshobbs) [inline docs/development/plugin-triggers.md]: typo change: "a more detailed output"
code block: `PLUGIN_NAME:help`
typo: "consistency"

**Comment by u2mejc** [comment]
@michaelshobbs 
```diff
-detail-
+detailed+
-PLUGIN_NAME:help-
+`PLUGIN_NAME:help`+
-consitancty-
+consistency+
```

**Comment by u2mejc** [comment]
:shipit: :shipit: :shipit:

**Comment by josegonzalez** [comment]
:+1:

**Comment by josegonzalez** [comment]
:clap:
